### DISCUSSION TITLE
Data in Traces

### DISCUSSION THREAD

**Discussion by ChristopherThomasHill** [ChampSim/ChampSim#477] [Ideas]
Is there any plan to incorporate data into the champsim simulator?

A feature like this would require many modifications including:
- A new trace format (see #382) that incorporate data
- Modifications to the channel and cache to allow movement of data of block_size length
- Modifications to ooo_cpu to allow misaligned memory accesses
- Modifications to DRAM to keep track of data being written/read
- Modifications to ptw  to support new data size

Currently I have a new trace format (integrated with the intel pin tool) that:
- Includes a serialized header that will track the initial state of pages (makes dynamically changing champsim page_size harder). This is to help allow prefetchers to prefetch the correct data.
- Serializes each instruction instead of being constant size. This helps since each instruction now may read or write some unique size of data.

**Comment by maccoymerrell** [comment]
I believe there is already some infrastructure in place within ChampSim for this. I understand there may be some advantages to knowing data within the core but tracking reads/writes in the DRAM seems like a lot of trouble. Massive increases in memory usage by ChampSim is one problem, but I can also imagine the performance impact being severe due to the additional cache stress. Not to mention deallocations of physical memory aren't simulated by ChampSim or in the traces (which can be fixed), so you can always guarantee ChampSim will use more memory simulating a process than the process itself took. 

Assuming there isn't anything outside the core that needs exact data to simulate, I think you could leverage the trace-based nature of ChampSim to do data-based simulation only in the core with far less overhead. If full data simulation is the goal, I feel like the memory hurdles would be too significant to add it as a feature. I have had ChampSim batch jobs that would be severely throttled if every instance needed 8GB+ of memory to run and would provide no benefits to those runs.

**Reply by ChristopherThomasHill** [reply]
You make a good point that the overhead might get too high to propagate and store it all the way in DRAM. Do you think it would be possible to include memory within some levels of the cache hierarchy? You suggested that the additional cache stress would impact performance, but is there any analysis on this?

**Reply by maccoymerrell** [reply]
It might be possible to include the data in the cache hierarchy, but you would need to prepopulate the data for all reads that go into the cache, since evictions/compulsory misses are going to happen. At that point you are just giving cache lines a data value, but not really using it for anything. Everything is going to need to come from the trace itself since dirty cache evictions effectively get erased. It would make the data that is in a cache line available within the simulation, but pretty much unused.

I don't think analysis is necessary to determine that cache stress is going to be an issue. Just to be clear, the cache stress I am referring to is of the machine running ChampSim, not the machine simulated by ChampSim. Suppose we did this and ChampSim's data footprint goes from a few hundred MBs to a few GBs. In that case, the performance will suffer since ChampSim is now having to share memory and cache space with the workload its trying to simulate. I think doing just the caches is far more feasible, since caches are much smaller than DRAM.

I don't think I could give an accurate number as to what magnitude the slowdown would be. What can be guaranteed is that running multiple ChampSim instances on a single node will become a lot more difficult due to the memory usage, and performance will suffer a lot from that. Having run ChampSim on congested machines before, if memory is thrashing a lot runtimes can increase 2x or 3x pretty easily.

**Reply by ngober** [reply]
Maccoy is absolutely correct here. I'll add that my own unofficial benchmarking analysis of ChampSim has shown that the model is actually rather cache unfriendly, arising from the way that it moves through each operable sequentially making its working set quite large. Any additional data being stored and passed around should be considered very carefully, especially when the data in question is on the order of 64 bytes per request.
